### PR TITLE
added new features filtering on computer names

### PR DESCRIPTION
**Pull request by hitenkoku**
## What Changed

- added `--include-computer` and '--exclude-computer' option in 'csv-timeline', `json-timeline`, `metrics`, `logon-summary`, `pivot-keywords-list` command

I would appreciate it if you could review when you have time.

### CODE REVIEW

**Comment by hitenkoku** [comment]
##Evidence

- conflict check

```
 ./1117.exe csv-timeline -d ..\hayabusa-sample-evtx\ -o 1117-include-computer.csv --include-computer IE8Win7 --exclude-computer IE8WIN11 -C -q
error: the argument '--include-computer <COMPUTER>' cannot be used with '--exclude-computer <COMPUTER>'
```

- help output
  - csv-timeline

```
> ./1117.exe help csv-timeline
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)

Usage:
  hayabusa.exe csv-timeline <INPUT> [OPTIONS]

...
Filtering:
  -E, --EID-filter                   Scan only common EIDs for faster speed (./rules/config/target_event_IDs.txt)
  -D, --enable-deprecated-rules      Enable rules with status of deprecated
  -n, --enable-noisy-rules           Enable rules set to noisy (./rules/config/noisy_rules.txt)
  -u, --enable-unsupported-rules     Enable rules with status of unsupported
  -e, --exact-level <LEVEL>          Scan for only specific levels (informational, low, medium, high, critical)
      --exclude-category <CATEGORY>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --exclude-computer <COMPUTER>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --exclude-status <STATUS>      Ignore rules according to status (ex: experimental) (ex: stable,test)
      --exclude-tags <TAGS>          Exclude load rules with specific tags (ex: sysmon)
      --include-category <CATEGORY>  Only load rules with certain logsource categories (ex: process_creation,pipe_created)
      --include-computer <COMPUTER>  Scan only these computer names (ex: ComputerA) (ex: ComputerA,ComputerB)
      --include-tags <TAGS>          Only load rules with specific tags (ex: attack.execution,attack.discovery)
  -m, --min-level <LEVEL>            Minimum level for rules (default: informational)
  -P, --proven-rules                 Scan only proven rule for faster speed (./rules/config/proven_rules.txt)
      --timeline-end <DATE>          End time of the event logs to load (ex: "2022-02-22 23:59:59 +09:00")
      --timeline-start <DATE>        Start time of the event logs to load (ex: "2020-02-22 00:00:00 +09:00")
...
```

  - json-timeline

```
> ./1117.exe help json-timeline
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)

Usage:
  hayabusa.exe json-timeline <INPUT> [OPTIONS]

...
Filtering:
  -E, --EID-filter                   Scan only common EIDs for faster speed (./rules/config/target_event_IDs.txt)
  -D, --enable-deprecated-rules      Enable rules with status of deprecated
  -n, --enable-noisy-rules           Enable rules set to noisy (./rules/config/noisy_rules.txt)
  -u, --enable-unsupported-rules     Enable rules with status of unsupported
  -e, --exact-level <LEVEL>          Scan for only specific levels (informational, low, medium, high, critical)
      --exclude-category <CATEGORY>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --exclude-computer <COMPUTER>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --exclude-status <STATUS>      Ignore rules according to status (ex: experimental) (ex: stable,test)
      --exclude-tags <TAGS>          Exclude load rules with specific tags (ex: sysmon)
      --include-category <CATEGORY>  Only load rules with certain logsource categories (ex: process_creation,pipe_created)
      --include-computer <COMPUTER>  Scan only these computer names (ex: ComputerA) (ex: ComputerA,ComputerB)
      --include-tags <TAGS>          Only load rules with specific tags (ex: attack.execution,attack.discovery)
  -m, --min-level <LEVEL>            Minimum level for rules (default: informational)
  -P, --proven-rules                 Scan only proven rule for faster speed (./rules/config/proven_rules.txt)
      --timeline-end <DATE>          End time of the event logs to load (ex: "2022-02-22 23:59:59 +09:00")
      --timeline-start <DATE>        Start time of the event logs to load (ex: "2020-02-22 00:00:00 +09:00")
...
```

  - metrics

```
> ./1117.exe help metrics                                                                                           ─╯
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)

Usage:
  hayabusa.exe metrics <INPUT> [OPTIONS]

...
Filtering:
      --exclude-computer <COMPUTER>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --include-computer <COMPUTER>  Scan only these computer names (ex: ComputerA) (ex: ComputerA,ComputerB)
...
```

  - logon-summary

```
./1117.exe help logon-summary
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)

Usage:
  hayabusa.exe logon-summary <INPUT> [OPTIONS]

...
Filtering:
      --exclude-computer <COMPUTER>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --include-computer <COMPUTER>  Scan only these computer names (ex: ComputerA) (ex: ComputerA,ComputerB)
...
```

  - pivot-keywords-list

```
> ./1117.exe help pivot-keywords-list                                                                               ─╯
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)

Usage:
  hayabusa.exe pivot-keywords-list <INPUT> [OPTIONS]

...
Filtering:
  -E, --EID-filter                   Scan only common EIDs for faster speed (./rules/config/target_event_IDs.txt)
  -D, --enable-deprecated-rules      Enable rules with status of deprecated
  -n, --enable-noisy-rules           Enable rules set to noisy (./rules/config/noisy_rules.txt)
  -u, --enable-unsupported-rules     Enable rules with status of unsupported
  -e, --exact-level <LEVEL>          Scan for only specific levels (informational, low, medium, high, critical)
      --exclude-computer <COMPUTER>  Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
      --exclude-status <STATUS>      Ignore rules according to status (ex: experimental) (ex: stable,test)
      --include-computer <COMPUTER>  Scan only these computer names (ex: ComputerA) (ex: ComputerA,ComputerB)
  -m, --min-level <LEVEL>            Minimum level for rules (default: informational)
      --timeline-end <DATE>          End time of the event logs to load (ex: "2022-02-22 23:59:59 +09:00")
      --timeline-start <DATE>        Start time of the event logs to load (ex: "2020-02-22 00:00:00 +09:00")

...
```

**Review by fukusuket — CHANGES_REQUESTED** [review]

**Comment by fukusuket** [comment]
@hitenkoku 
The following commands have the same result.  I would appreciate it if you could check it 🙇
```
$ ./hayabusa csv-timeline -d ../hayabusa-sample-evtx --include-computer A
$ ./hayabusa csv-timeline -d ../hayabusa-sample-evtx --include-computer B
...

Timestamp ‖ Computer ‖ Channel ‖ EventID ‖ Level ‖ RecordID ‖ RuleTitle ‖ Details ‖ ExtraFieldInfo
2016-09-20 01:50:06.513 +09:00 ‖ - ‖ - ‖ - ‖ med ‖ ‖ PW Guessing ‖ [condition] count() by IpAddress >= 5 in timeframe [result] count:3558 IpAddress:192.168.198.149 timeframe:5m ‖
2019-05-01 04:27:02.847 +09:00 ‖ - ‖ - ‖ - ‖ med ‖ ‖ PW Spray ‖ [condition] count(TargetUserName) by IpAddress >= 5 in timeframe [result] count:41 TargetUserName:cragoso/ebooth/cfleener/tbennett/sanson/sarmstrong/melliott/jleytevidal/lpesce/wstrzelec/edygert/thessman/jlake/econrad/celgee/zmathis/jorchilles/mtoussain/cspizor/bking/Administrator/drook/jkulikowski/dpendolino/bhostetler/bgalbraith/mdouglas/lschifano/cmoody/dmashburn/kperryman/jwright/eskoudis/smisenar/rbowes/ssims/baker/bgreenwood/cdavis/gsalinas/psmith IpAddress:172.16.144.128 timeframe:5m ‖
2019-05-01 04:32:03.525 +09:00 ‖ - ‖ - ‖ - ‖ med ‖ ‖ PW Spray ‖ [condition] count(TargetUserName) by IpAddress >= 5 in timeframe [result] count:14 TargetUserName:cragoso/edygert/jlake/jorchilles/cspizor/bking/drook/bgalbraith/mdouglas/dmashburn/ssims/smisenar/baker/bgreenwood IpAddress:172.16.144.128 timeframe:5m ‖
...
Top 5 computers with most unique detections:
critical: n/a
high: n/a
medium: n/a
low: n/a
informational: n/a
...
```
It seems that the detection rule that hits the `aggregate condition` is not filtered by computername? 🤔

File: src/detections/configs.rs [hunk 1]
```diff
@@ -693,6 +789,14 @@ pub struct DetectCommonOption {
     /// Output verbose information
     #[arg(help_heading = Some("Display Settings"), short = 'v', long, display_order = 480)]
     pub verbose: bool,
+
+    /// Scan only these computer names (ex: ComputerA) (ex: ComputerA,ComputerB)
+    #[arg(help_heading = Some("Filtering"), long = "include-computer", value_name = "COMPUTER", conflicts_with = "exclude-computer", use_value_delimiter = true, value_delimiter = ',', display_order = 351)]
+    pub include_computer: Option<Vec<String>>,
+
+    /// Do not load rules with certain logsource categories (ex: process_creation,pipe_created)
```
[the remark this answers is not in the document]
Comment (hitenkoku) [reply src/detections/configs.rs:797]: Thanks for your suggestion. I adapt it

**Comment by hitenkoku** [comment]
@fukusuket 

The `aggregation condition` may be aggregated across multiple terminals, and originally the name of the Computer in the output was also `-`.


Filtering for `aggregation condition` is also time-consuming.

**Comment by YamatoSecurity** [comment]
@hitenkoku Thank you very much for the implementation!
Do these new options filter for computer name after scanning or before scanning the event? If we are only filtering on computer name after scanning, then there will be no speed/memory improvements, correct? I want to check the computer name before any rule matching occurs to improve performance as much as possible. When the count aggregator is checking the event information, it should be able to check the computer name, no?

**Comment by hitenkoku** [comment]
@YamatoSecurity Thank you for your comment.
We are now filtering after detection. I will try to implement it to see if we can filter before detection.

**Comment by hitenkoku** [comment]
@fukusuket @YamatoSecurity Computerized filtering before detection at https://github.com/Yamato-Security/hayabusa/pull/1124/commits/24516b2d4b34ec124121af6337c05bc61fdcb61a The following modifications have been made to the system.

Sorry for the repetition. Would you please check it out?

**Review by fukusuket — APPROVED** [review]
I confirmed as follows,  LGTM!!🚀

**csv-timeline/json-timeline**
- No unnecessary diffs compared to After merging the results of `--include-computer` and `--exclude-computer`

**metrics**
- I confirmed that the results of the `search` command and the total number of the `metrics` command match.  
```
% ./hayabusa metrics -d ../hayabusa-sample-evtx --include-computer IE8Win7 -o metrics.csv
% awk -F"," '{a+=$1} END{print a;}' metrics.csv
8603

% ./hayabusa search -d ../hayabusa-sample-evtx -r ".*" -F Computer:"IE8Win7" -o search.csv
...
Total findings: 8603
```

**Review by YamatoSecurity — APPROVED** [review]
@hitenkoku LGTM! Thanks so much!
Scans are now much faster when filtering on computers.

**Pull request merged**